### DISCUSSION TITLE
Weird difference in waveforms drawn based on Peaks provided or not

### DISCUSSION THREAD

**Discussion by Stianlars1** [katspaugh/wavesurfer.js#3191] [Q&A]
FYI: 
-  I've upgraded to v7.1.3, hoping this would be A solution, but it wasn't.
- Im working in React v^18.2.0
- MacOS.

**Lets just dive right into it, shall we?**

i have updated my private music website to enable Typescript support, and also caching of **peaks** in the local Storage for faster drawings on load. The **problem** i'm having is that the drawings of the waveforms are different when im loading the instance of wavesurfers with peaks, which means that my peaks looks different when i load it without peaks. The problem is that the drawings of the waveforms looks much nicer and more accurate for the instance loaded without peaks, than the one with peaks provided.

**images to show the difference**
This image shows the waveform drawn by **providing** peaks!
![Skjermbilde 2023-09-14 kl  08 56 54](https://github.com/katspaugh/wavesurfer.js/assets/71408982/75c57dd1-8449-4f26-bfc2-edb100eb03c6)

While this picture shows the waveforms drawn **WITHOUT** providing peaks.
![Skjermbilde 2023-09-14 kl  08 57 13](https://github.com/katspaugh/wavesurfer.js/assets/71408982/9e3306b4-27ca-4c38-8fc7-ab27dc0cde85)

As you can se theres a big difference in how they look and appear, and i have tried several things, like changing how the exportPeaks works, passed in different options, but without luck. Now im asking you to please help me.

**Lets take a look at the code.**

**For handling the peaks ETC.. the code below should give you an idea of how i do things:**

_AudioPlayer.tsx_
```
// Getting the peaks if theyre stored:
  const peaksFromLocalStorage = localStorage.getItem(`PEAKS_${identifier}`)
  const peaks: Peaks = peaksFromLocalStorage ? JSON.parse(peaksFromLocalStorage) : null
  const wavesurfer = useWavesurfer(containerRef, props, peaks)

// in the useEffect i have:
  wavesurfer.on('decode', () => {
      if (peaks) {
          // check if peaks has changed
          const newPeaks = wavesurfer.exportPeaks()
          // checking if the peaks i have cached are the same
          const equally = areEqual(peaks, newPeaks)
          if (!equally) {
              return localStorage.setItem(`PEAKS_${identifier}`, JSON.stringify(newPeaks))
          }
      } 
    // If no peaks are stored, then add them to local Storage
    if (identifier) {
              const peaks = wavesurfer.exportPeaks()
              localStorage.setItem(`PEAKS_${identifier}`, JSON.stringify(peaks))
          }
  })
```

Then what i do in my actual creation of the wavesurfers is to check wheter i have peaks or not, and then creating the wavesurfer instance based on that, like this: 

_useWavesurfer.tsx_

```
export const useWavesurfer = (containerRef: any, options: AudioPlayerProp, peaks: Peaks) => {
    const [wavesurfer, setWavesurfer] = useState<WaveSurfer>()

    let ws: WaveSurfer
    useEffect(() => {
        if (!containerRef.current) return
        if (peaks) {
            console.log('PEAKS EXISTED!!')
            ws = WaveSurfer.create({
                ...options,
                container: containerRef.current,
                peaks: peaks,
            })
        } else if (!peaks) {
            console.log('NO PEAKS :((((!')
            ws = WaveSurfer.create({ ...options, container: containerRef.current })
        }

        setWavesurfer(ws)

        return () => {
            ws.destroy()
            setWavesurfer(undefined)
        }
    }, [options, containerRef])

    return wavesurfer
}

```

Let me know if you need any more info regarding the problem.

**EDIT**
After looking into the answer provided in the comments i have modified the exportPeaks to this: 
```
                    const peaks = wavesurfer.exportPeaks({
                        channels: 2,
                        precision: 10_000,
                        maxLength: 8000,
                    })
```
From that i get these waveforms drawn: 
![Skjermbilde 2023-09-14 kl  09 10 33](https://github.com/katspaugh/wavesurfer.js/assets/71408982/84fe6b3e-25c0-49e8-af25-674b1ecdd2a0)
As you can see they are exactly like the first picture where i provided peaks in the first place. 
Result: No changes to the waveforms.

**Comment by katspaugh** [comment]
`exportPeaks` takes three named parameters that control the "resolution" at which the peaks are exported.

```typescript
  public exportPeaks({ channels = 1, maxLength = 8000, precision = 10_000 } = {}): Array<number[]> {
```

By default it will apply some sampling and rounding, and it will only export one channel. You might try passing highrer numbers for these params and see if it makes it look better.

Alternatively, you can use `wavesurfer.getDecodedData()` to get the original peaks (as an AudioBuffer), and apply your own serialization for it to be suitable for localStorage.

Edit:

Looking at your screenshot, the main issue seems to be that it becomes mono. You can pass `channels: 2` to export both channels.

**Reply by Stianlars1** [reply]
i just updated my whole question for you! please read if if you havent already 🔥

**Reply by Stianlars1** [reply]
If i export the peaks with 2 channels, it will give me 2 arrays instead of 1, should i just provide the [[1], [2]] to the creation of the wavesurfers instance like that? peaks: [[firstChannel], [secondChannel]]?

Let me do some modifications to this, and get back to you!

**Reply by katspaugh** [reply]
Yes, exactly.

**Reply by Stianlars1** [reply]
I have edited my post to include a new screenshot and an "edited" section at the bottom. Please look at it

**Reply by Stianlars1** [reply]
Ok, an update for how it went. I took your suggestion on using the `wavesurfer.getDecodedData()` which gives me a whole other data than the `exportPeaks()` function. The problem here is that the new data i receive is to large for localstorage which led me to create a server for posting the peaks to a backend. This seems to take longer time than just waiting for the generation of the peaks, so im right back where i started. 


Do you know if there is possible to load the peaks **only** while the wavesurfer instance is loading the audio and generating waveforms,a nd then when its loaded and done drawing i can remove the peaks i added and then see the generated waveform without using peaks? if you understand what i mean.

**Reply by katspaugh** [reply]
Have you tried increasing the `precision` and `maxLength` parameters?

**Reply by Stianlars1** [reply]
Yeah! It could be that I would need to use the getDecodedData() function to get the original data peaks. 

Right now I can't really test the original peaks since I have no place to store them. 

What would you recommend me do to save the/cache the peaks? I have looked into server-side storing, but the peaks stringified seems to be about 100+ MB which is not durable. 

Any tips on storing peaks?

**Reply by katspaugh** [reply]
I wouldn’t recommend storing the original audio buffer. It’s essentially an entire uncompressed audio file.

**Reply by Stianlars1** [reply]
So you're suggesting using the exportPeaks data and store those? 

Is there a way for me to fix the weird difference I have in the waveforms then?

**Reply by katspaugh** [reply]
You could do something like this, it's the same max-peak sampling algorithm that wavesurfer uses when rendering:

```js
wavesurfer.once('ready', () => {
  const maxLength = 4000
  const buffer = wavesurfer.getDecodedData()
  const channels = Math.min(2, buffer.numberOfChannels)
  const peaks = []
  for (let i = 0; i < channels; i++) {
    const channel = buffer.getChannelData(i)
    const data = []
    const sampleSize = Math.round(channel.length / maxLength)
    for (let i = 0; i < channel.length; i += sampleSize) {
      const sample = channel.slice(i, i + sampleSize)
      data.push(Math.max.apply(Math, sample))
    }
    peaks.push(data)
  }

  // Final peaks
  console.log(peaks)
})
```

**Reply by Stianlars1** [reply]
Thank you so much for providing this. Will check it out later tonight. 

I may sound dumb, but how can I "use" those peaks or draw them for that instance after the audio has been loaded? Let's say I have the DATA array with the peaks, how to I load them to that canvas? 

You see, I'm just familiar with providing the peaks as options when I create the instance. I had no idea we could provide them to a created instance later on?

**Reply by katspaugh** [reply]
Drawing them on the same instance wouldn't make any difference. You're supposed to save them into localStorage as a form of cache as per your original request.
